polar2grid/writers/hdf5.py

```diff
@@ -0,0 +1,293 @@
+
+import os
+import sys
+import logging
+
+import satpy
+import h5py
+import numpy as np
+import xarray as xr
+import dask.array as da
+
+from itertools import groupby
+from datetime import datetime as datetime
+from satpy.dataset import DataID
+from satpy.writers import ImageWriter
+
+from trollsift import Parser
+
+from pyresample.geometry import SwathDefinition
+from polar2grid.utils.legacy_compat import convert_p2g_pattern_to_satpy
+from polar2grid.writers.geotiff import NUMPY_DTYPE_STRS, NumpyDtypeList, str_to_dtype, int_or_float
+
+from polar2grid.core.script_utils import setup_logging, rename_log_file, create_exc_handler
+
+USE_POLAR2GRID_DEFAULTS = bool(int(os.environ.setdefault("USE_POLAR2GRID_DEFAULTS", "1")))
+
+LOG = logging.getLogger(__name__)
+
+# reader_name -> filename
+DEFAULT_OUTPUT_FILENAMES = {None: "{platform_name}_{sensor}_{start_time:%Y%m%d_%H%M%S}.h5"}
+
+
+def all_equal(iterable: list[str]) -> bool:
+    "Returns True if all the elements are equal to each other"
+    g = groupby(iterable)
+    return next(g, True) and not next(g, False)
+
+
+class FakeHDF5:
+    def __init__(self, output_filename: str, var_name: str):
+        self.output_filename = output_filename
+        self.var_name = var_name
+
+    def __setitem__(self, write_slice, data):
+        with h5py.File(self.output_filename, mode="a") as fh:
+            fh[self.var_name][write_slice] = data
+
+
+class hdf5writer(ImageWriter):
+    """Writer for hdf5 files."""
+
+    def __init__(self, **kwargs):
+        """Init the writer."""
+        super(hdf5writer, self).__init__(**kwargs)
+
+    def _output_file_kwargs(self, dataset, dtype):
+        """Get file keywords from data for output_pattern."""
+        if isinstance(dataset, list):
+            dataset = dataset[0]
+            area = dataset[0].attrs["area"]
+        else:
+            area = dataset.attrs["area"]
+
+        args = dataset.attrs
+        args["grid_name"] = "native" if isinstance(area, SwathDefinition) else area.area_id
+        args["rows"], args["columns"] = area.shape
+        args["data_type"] = dtype
+
+        return args
+
+    def iter_by_area(self, datasets: list[xr.DataArray]):
+        """Generate datasets grouped by Area.
+        :return: list of (area_obj, list of dataset objects)
+        """
+        datasets_by_area = {}
+        for ds in datasets:
+            a = ds.attrs.get("area")
+            datasets_by_area.setdefault(a, []).append(ds)
+        return datasets_by_area.items()
+
+    def get_filename(self, **kwargs) -> str:
+        """Create a filename for saving output data.
+
+        Args:
+            kwargs (dict): Attributes and other metadata to use for formatting
+                the previously provided `filename`.
+
+        """
+        if self.filename_parser is None:
+            raise RuntimeError("No filename pattern or specific filename provided")
+        self.filename_parser = Parser(convert_p2g_pattern_to_satpy(self.filename_parser.fmt))
```

Comment:
Could this conversion happen in the `__init__` of the writer? That way you don't have to subclass this whole method, right?

Or, if we assume that all patterns configured for the writer are current/correct, then we could do the conversion on the `--output-filename` command line argument. Thoughts?

Reply:
Are you referring converting in glue at _parse_writer_args? (Assuming all patterns configured are correct).

Reply:
I was more thinking of specifying the `type` in the `add_argument` to be a function that does the conversion. You can see this idea in the `int_or_float` usage in the command line arguments for your writer already (or at least `--fill-value` was using it).